### ISSUE TITLE
Zeebe exporting latency can increase significantly without clear root cause

### ISSUE THREAD

**Issue by Zelldon** [camunda/camunda#21376]
**Describe the bug**

I have observed several times in my recent benchmarks that the exporting latency, the time from writing a record to start exporting the record, has increased significantly. See new metric [PR](https://github.com/camunda/camunda/pull/21297)

Normally it is expected that this latency is similar to the processing around some milliseconds, but on higher load benchmarks it has been seen that this latency can increase to minutes.

![2024-08-20_14-11](https://github.com/user-attachments/assets/03653b57-f9a5-4797-820a-7fc6da07c93c)

This obviously also impacts the importing latency and import throughput.

![2024-08-20_13-28](https://github.com/user-attachments/assets/390f6307-6199-41c9-9854-7eaf0175f424)

Right now it is not clear whether the impact is due to ES and the Elasticsearch exporter flush duration or something else.

During the benchmark, several metrics have been investigated. For example, no pods have been restarted, the disk usage was around 30% (for ES).

The only interesting was that a ES seem to create a new instance around 11, when the latency was degraded. 

![2024-08-20_14-55](https://github.com/user-attachments/assets/39aaba78-04d6-45a2-882a-361bb8a06312)

Furthermore, in [ES logs I can see that some queries take quite a while. 
](https://console.cloud.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.cluster_name%3D%22zeebe-cluster%22%0Aresource.labels.namespace_name%3D%22ck-20-08-high-load-base%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fcomponent%3D%22master%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Finstance%3D%22zeebe-benchmark-test%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fname%3D%22elasticsearch%22;pinnedLogId=2024-08-20T08:59:30.135105502Z%2Favew5v3u11wpuiib;cursorTimestamp=2024-08-20T08:59:33.341274202Z;startTime=2024-08-20T08:56:00.000Z;endTime=2024-08-20T09:08:00.000Z?project=zeebe-io)

![2024-08-20_14-56](https://github.com/user-attachments/assets/61119fe6-f6c6-437e-9841-b160b7f26250)

**To Reproduce**

Set up a benchmark with a higher load, with Operate enabled. It has happened at least in my two recent benchmarks.

<details>
<summary>Lower flush delay with high load: Helm install command</summary>
<pre>
helm install $(releaseName) $(chartPath) --render-subchart-notes \
 --set global.image.tag=ck-operate-benchmark-1ad8f375 \
 --set camunda-platform.zeebe.image.repository=gcr.io/zeebe-io/zeebe \
 --set camunda-platform.zeebe.image.tag=ck-operate-benchmark-1ad8f375 \
 --set camunda-platform.zeebeGateway.image.repository=gcr.io/zeebe-io/zeebe \
 --set camunda-platform.zeebeGateway.image.tag=ck-operate-benchmark-1ad8f375 \
 --set starter.rate=50 \
 --set worker.replicas=1 \
 --set timer.replicas=1 \
 --set timer.rate=50 \
 --set publisher.replicas=1 \
 --set publisher.rate=50 \
 --set camunda-platform.operate.enabled=true \
 --set camunda-platform.operate.image.repository=gcr.io/zeebe-io/operate \
 --set camunda-platform.operate.image.tag=ck-operate-benchmark \
 --set camunda-platform.elasticsearch.master.persistence.size=128Gi \
 --set camunda-platform.zeebe.retention.minimumAge=1d \
 --set zeebe.config.zeebe.broker.exporters.elasticsearch.args.bulk.delay=1
</pre>
</details>

**Expected behavior**

The latency is not increasing.

**Impact**

The impact is performance-related, but also usability, as Operate becomes useless at some point, if it takes ages to load new data. This is why I mark the impact/severity as high. The likelihood seems to be high as well, as it happened on all my recent benchmarks.

**Comment by Zelldon** [comment]
Thanks to @deepthidevaki and @lenaschoenburg I looked into ES metrics, and whether ES is maybe overloaded.

I improved my benchmark dashboard to better understand the scenario that caused this.

![elastic-cpu-usage](https://github.com/user-attachments/assets/5a907eac-7023-477d-aec3-0ca5468e0159)

We can see that every hour a new index is created. This is because of this option, set in the benchmarks https://github.com/zeebe-io/benchmark-helm/blob/main/charts/zeebe-benchmark/values.yaml#L94

When creating a new index, this might move traffic/load to a different node.

@lenaschoenburg pointed out that the CPU used in the benchmarks might be limited not to 2, but already capped before as we use 

```
    resources:
      requests:
        ## @param elasticsearch.master.resources.requests.cpu cpu request
        cpu: 1
        ## @param elasticsearch.master.resources.requests.memory request
        memory: 2Gi
      limits:
        ## @param elasticsearch.master.resources.limits.cpu cpu limit
        cpu: 2
        ## @param elasticsearch.master.resources.limits.memory memory limit
        memory: 2Gi
```
Source: 
https://github.com/camunda/camunda-platform-helm/blob/main/charts/camunda-platform-latest/values.yaml#L3424-L3434

This means the pod is scheduled to a node, where still 1 CPU is feasible. It can take up to 2 CPUs, but this is not guaranteed. It can be that it is capped before, as other pods might use the CPU time. Priority of CPU scheduling is based on high requests, see [docs 
](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#how-pods-with-resource-requests-are-scheduled:~:text=The%20CPU%20request%20typically%20defines%20a%20weighting.%20If%20several%20different%20containers%20(cgroups)%20want%20to%20run%20on%20a%20contended%20system%2C%20workloads%20with%20larger%20CPU%20requests%20are%20allocated%20more%20CPU%20time%20than%20workloads%20with%20small%20requests.)

What we can see on the dashboard above is that before 10 am. Elastic-0 handled most traffic, the rest was divided by ES-1 and ES-2. 

After 10 it was more evenly distributed. 
After 11, when a new index was created again, the workload was again distributed. ES-0 received again most work again but ran on its limits of around 170% CPU time. Furthermore, ES-1 received a load which caused it to still run close to 100% CPU time. ES-2 received quite little traffic, and not really using all its potential.

In that sense, it seems to be bad scheduling of nodes and then also indexes can cause exporting delays, which causes a growing backlog. This in consequence is then causing more delay in the importing.


In order to prove this I'm currently running a benchmark where request and limit are set to 2 CPU, to verify whether we run into the same issue.

**Comment by Zelldon** [comment]
I have run a higher load experiment with request and limit 2 core for elastic search:

<details>
<summary>Lower flush delay: Helm install command</summary>
<pre>
helm install $(releaseName) $(chartPath) --render-subchart-notes \
 --set global.image.tag=ck-operate-benchmark-1ad8f375 \
 --set camunda-platform.zeebe.image.repository=gcr.io/zeebe-io/zeebe \
 --set camunda-platform.zeebe.image.tag=ck-operate-benchmark-1ad8f375 \
 --set camunda-platform.zeebeGateway.image.repository=gcr.io/zeebe-io/zeebe \
 --set camunda-platform.zeebeGateway.image.tag=ck-operate-benchmark-1ad8f375 \
 --set starter.rate=5 \
 --set worker.replicas=1 \
 --set timer.replicas=1 \
 --set timer.rate=5 \
 --set publisher.replicas=1 \
 --set publisher.rate=5 \
 --set camunda-platform.operate.enabled=true \
 --set camunda-platform.operate.image.repository=gcr.io/zeebe-io/operate \
 --set camunda-platform.operate.image.tag=ck-operate-benchmark \
 --set camunda-platform.elasticsearch.master.persistence.size=128Gi \
 --set camunda-platform.elasticsearch.master.resources.requests.cpu=2 \
 --set camunda-platform.elasticsearch.master.resources.limits.cpu=2 \
 --set camunda-platform.zeebe.retention.minimumAge=1d
</pre>
</details>

We can see that in general, it went smoother until it ran out of disk. We had a bigger spike in the importing latency, but it was able to recover.

![higher-export-flush-causing-import-delay](https://github.com/user-attachments/assets/84979fcf-eb7c-406d-bf14-512f7e7dac24)

I investigated a bit further why this spike has happened. We can see that the flush duration has increased way before, causing to increase the exporting latency and some point significantly the importing latency.


![elastic-cpu-throttling](https://github.com/user-attachments/assets/28bad15d-1c81-41f7-b617-41fda896c301)

As already mentioned before, every hour a new index are created causing to shift in the load of different nodes (that is because of our configuration in the benchmarks). Depending on the node where the pod is scheduled it has more or less CPU available (even if the request and limit are defined as 2 cores). 

We can see based on the metrics that for example elastic-2 fights between 17 and 20 with high CPU throttling. Furthermore, Elastic-2 seems to consume more CPU than others even if it has during the time not have that much indexing to do (likely other tasks in the background, we can't see).

Around 19:00 we can see that the throttling of the CPU is decreasing, and also elastic-1 takes more work due to new indexes. This allows us to decrease the import latency again.

With that, I will conclude that the exporting and of course importing (into Operate indices) latency is highly dependent on how Elasticsearch is performing (and consuming resources). We need to take care of related Elasticsearch metrics as well to understand certain side-effects and depending on the scenario give enough resources to ES. 

This means that the exporting latency, meaning the time from write-to-exporting latency can be affected negatively by a bad-performing sub-system like Elastic, as right now the Elasticsearch exporter is synchronously flushing. The same also applies of course when any other exporter performs badly. This we should consider also when implementing https://github.com/camunda/camunda/issues/16912 


P.S.: For reference the benchmark dashboard I used:

[Camunda Benchmarks - CK-1723726136320.json](https://github.com/user-attachments/files/16687548/Camunda.Benchmarks.-.CK-1723726136320.json)
